src/main/java/io/numaproj/numaflow/batchmapper/Service.java

```diff
@@ -0,0 +1,163 @@
+
+    // batchMapTaskExecutor is the executor for the batchMap. It is a fixed size thread pool
+    // with the number of threads equal to the number of cores on the machine times 2.
+    private final ExecutorService batchMapTaskExecutor = Executors
+            .newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
+
+    // SHUTDOWN_TIME is the time to wait for the sinker to shut down, in seconds.
+    // We use 30 seconds as the default value because it provides a balance between giving tasks enough time to complete
+    // and not delaying program termination unduly.
+    private final long SHUTDOWN_TIME = 30;
+
+    private final BatchMapper batchMapper;
+
+    @Override
+    public StreamObserver<Batchmap.BatchMapRequest> batchMapFn(StreamObserver<Batchmap.BatchMapResponse> responseObserver) {
+
+        if (this.batchMapper == null) {
+            return io.grpc.stub.ServerCalls.asyncUnimplementedStreamingCall(
+                    BatchMapGrpc.getBatchMapFnMethod(),
+                    responseObserver);
+        }
+
+        DatumIteratorImpl datumStream = new DatumIteratorImpl();
+
+        Future<BatchResponses> result = batchMapTaskExecutor.submit(() -> this.batchMapper.processMessage(
+                datumStream));
+
+        return new StreamObserver<Batchmap.BatchMapRequest>() {
+            @Override
+            public void onNext(Batchmap.BatchMapRequest mapRequest) {
+                try {
+                    datumStream.writeMessage(constructHandlerDatum(mapRequest));
+                } catch (InterruptedException e) {
+                    Thread.interrupted();
+                    onError(e);
+                }
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                // We close the stream and let the sender retry the messages
+                log.error("Error Encountered in batchMap Stream", throwable);
+                var status = Status.UNKNOWN.withDescription(throwable.getMessage()).withCause(throwable);
+                responseObserver.onError(status.asException());
+            }
+
+            @Override
+            public void onCompleted() {
+                try {
+                    // We Fire off the call to the client from here and stream the response back
+                    datumStream.writeMessage(HandlerDatum.EOF_DATUM);
+                    BatchResponses responses = result.get();
+                    log.debug(
+                            "Finished the call Result size is :{} and iterator count is :{}",
+                            responses.getItems().size(),
+                            datumStream.getCount());
+                    // Crash if the number of responses from the users don't match the input requests ignoring the EOF message
+                    if (responses.getItems().size() != datumStream.getCount() - 1) {
+                        throw new RuntimeException("Number of results did not match");
```

Comment:
Consider adding the real numbers in the error message, for better debuggability.

Reply:
Done